tools/syz-kconf/kconf.go

```diff
@@ -0,0 +1,377 @@
+	outputFile := filepath.Join(ctx.ConfigDir, ctx.Inst.Name+".config")
+	if verifyErr != nil {
+		// Write bad config for debugging with .tmp suffix so that it's not checked-in accidentially.
+		outputFile += ".tmp"
+		if err := osutil.WriteFile(outputFile, original); err != nil {
+			return fmt.Errorf("failed to write output file: %v", err)
+		}
+		return fmt.Errorf("%vsaved config to %v (run oldconfig)", verifyErr, outputFile)
+	}
+	return osutil.WriteFile(outputFile, config)
+}
+
+func (ctx *Context) executeShell() error {
+	for _, shell := range ctx.Inst.Shell {
+		if !ctx.Inst.Features.Match(shell.Constraints) {
+			continue
+		}
+		args := strings.Split(shell.Cmd, " ")
+		if args[0] != "make" {
+			return fmt.Errorf("non-make shell is not supported yet")
+		}
+		if err := ctx.Make(args[1:]...); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ctx *Context) applyConfigs(cf *kconfig.ConfigFile) {
+	for _, cfg := range ctx.Inst.Configs {
+		if !ctx.Inst.Features.Match(cfg.Constraints) {
+			continue
+		}
+		if cfg.Value != kconfig.No {
+			// If this is a choice, first unset all other options.
+			// If we leave 2 choice options enabled, the last one will win.
+			// It can make sense to move this code to kconfig in some form,
+			// it's needed everywhere configs are changed.
+			if m := ctx.Kconf.Configs[cfg.Name]; m != nil && m.Parent.Kind == kconfig.MenuChoice {
+				for _, choice := range m.Parent.Elems {
+					cf.Unset(choice.Name)
+				}
+			}
+		}
+		cf.Set(cfg.Name, cfg.Value)
+	}
+}
+
+func (ctx *Context) verifyConfigs(cf *kconfig.ConfigFile) error {
+	errs := new(Errors)
+	for _, cfg := range ctx.Inst.Configs {
+		act := cf.Value(cfg.Name)
+		if act == cfg.Value || cfg.Optional || !ctx.Inst.Features.Match(cfg.Constraints) {
+			continue
+		}
+		if act == kconfig.No {
+			errs.push("%v:%v: %v is not present in the final config", cfg.File, cfg.Line, cfg.Name)
+		} else if cfg.Value == kconfig.No {
+			errs.push("%v:%v: %v is present in the final config", cfg.File, cfg.Line, cfg.Name)
+		} else {
+			errs.push("%v:%v: %v does not match final config %v vs %v",
+				cfg.File, cfg.Line, cfg.Name, cfg.Value, act)
+		}
+	}
+	return errs.err()
+}
+
+func (ctx *Context) addUSBConfigs(cf *kconfig.ConfigFile) error {
+	android := ""
+	if ctx.Inst.Features[featAndroid] {
+		android = "android"
+	}
+	// TODO: enable all USB configs for non-android instances (are there any bad/debug ones?).
+	// TODO: enable all configs from distro configs, not just USB (are there any interesting ones?).
+	distroConfig := filepath.Join(ctx.ConfigDir, "distros", android+"*")
+	// Some USB drivers don't depend on any USB related symbols, but rather on a generic symbol
+	// for some input subsystem (e.g. HID), so include it as well.
+	// TODO: don't include HID, it does not seem to add anything relevant.
+	return ctx.addDependentConfigs(cf, []string{"USB_SUPPORT", "HID"}, []string{"USB_GADGET"}, distroConfig)
```

Comment:
@xairy PTAL

You script also use USB_ARCH_HAS_HCD and USB. But these already depend on USB_SUPPORT, so don't seem to be necessary. Maybe the python library mishandled transitive dependencies?

I am also not sure why HID is need. If I remove it, I see only some irrelevant configs go away, e.g. ps/2 mouse. I will remove it in a separate commit.

Re these configs:
https://github.com/google/syzkaller/blob/ff4a3345a1b2a40ff1b8b983153d0b1fcc72f1c5/dashboard/config/util.sh#L47-L72
Did you want to remove everything that depends on USB_GADGET?

Reply:
> You script also use USB_ARCH_HAS_HCD and USB. But these already depend on USB_SUPPORT, so don't seem to be necessary.

This could be.

> I am also not sure why HID is need. If I remove it, I see only some irrelevant configs go away, e.g. ps/2 mouse. I will remove it in a separate commit.

AFAIR HID uses its own drivers dispatching system, and some of the HID drivers don't depend on USB directly. Maybe something works differently with your parser...

> Did you want to remove everything that depends on USB_GADGET?

No, the intention was to disable all legacy gadget drivers (see `drivers/usb/gadget/legacy/Kconfig`) expect for `USB_RAW_GADGET`.

[...]

Looking at the new USB config: `CONFIG_USB_DUMMY_HCD` and `CONFIG_USB_RAW_GADGET ` must be enabled.

Reply:
> AFAIR HID uses its own drivers dispatching system, and some of the HID drivers don't depend on USB directly. Maybe something works differently with your parser...

The new parser extracts dependencies transitively, however I think the old one did it as well.

I will prepare a change and we will look at the actual, then it will be easier to decide.
Maybe that's because I disabled everything that depends on USB_GADGET, thus I did not see anything useful that depends on HID.

Reply:
Fixed things.

CONFIG_USB_RAW_GADGET and CONFIG_USB_DUMMY_HCD are now enabled.
Also, we don't disable everything that depends on USB_GADGET, but rather disable the specific configs you disabled before.

However, there is still a significant diff. Some MEDIA/DVB configs are not enabled now, but they don't depend on USB, so I am not sure why they were enabled before...